### PR TITLE
Add chooseRenormalize utility function

### PR DESCRIPTION
**Pull request by psbanka**
**⚠️ Note this is a PR into the v.2.0.0 branch and not master**

Ports an internal python random selection algorithm to a  utility library. This will allow us to perform random task selection and bucketing based on the given number of available task buckets i.e. `max_tasks`.

- [X] Port basic Python over
- [X] Add 100% code-coverage (ensuring that the JS code works the same as the Python
- [X] Mock random-number generation for test consistency
- [X] Add typing for random-mock.

### CODE REVIEW

**Comment by psbanka** [comment]
I added a pared-down version of the Python code here: https://github.com/fastly/insights.js/pull/29/commits/ab84597f727d78721b763252c65bc5c10c0898bb#diff-9c38e7aa72e0f633a2b2bbb4e9af7fe8 so that I could compare the JS and the Python directly, if that's helpful.

File: src/util/chooseRenormalize.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  return values.reduce((memo: number, i: number) => memo + i) / values.length;
+}
+
+/**
+ * This defines a random number between 0 and "mass", which should cover the entire
+ * domain of the probability distribution. Then we will "walk" the probability mass
+ * until we exceed the generated value
+ */
+function basicChooseOne(probs: number[], mass: number): number {
+  const r = mass * Math.random();
```
Comment (rlanda) [inline src/util/chooseRenormalize.ts]: Where are we setting the seed for the random number generator?

File: src/util/chooseRenormalize.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      .fill(null)
+      .map((x: null, index: number) => index);
+
+  // Since we are mutating the probabilities vector, we copy it first
+  const p = probs.map(x => x);
+
+  // This assumes that the probability vector is normalized. If it is not, it
+  // should be changed to the sum of the weights
+  let remaining = 1;
+
```
Comment (rlanda) [inline src/util/chooseRenormalize.ts]: Are we actually making sure that the weights are normalized somewhere?

File: src/util/chooseRenormalize.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  }
+  return i;
+}
+
+function chooseRenormalize(probs: number[], k: number): number[] {
+  //  If we have more work slots than candidates in the probability distribution,
+  //  we return them all
+  if (k >= probs.length)
+    return Array(probs.length)
+      .fill(null)
```
Comment (phamann) [inline src/util/chooseRenormalize.ts]: It would be good to see the size of the core-js `Array.prototype.fill()` polyfill that babel preset env compiles for this, as it it may not be worth the overhead.

File: src/util/chooseRenormalize.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      .map((x: null, index: number) => index);
+
+  // Since we are mutating the probabilities vector, we copy it first
+  const p = probs.map(x => x);
```
Comment (phamann) [inline src/util/chooseRenormalize.ts]: Style nit: but I find the spread operator variant nicer than map.
```
const p = [...props]`;
```

File: .eslintrc [hunk 5]
```diff
@@ -38,6 +38,7 @@
         prefer-const: 1,
         no-const-assign: 2,
         "@typescript-eslint/no-explicit-any": 0,
+        "@typescript-eslint/explicit-function-return-type": 0,
```
Comment (phamann) [inline .eslintrc]: Would be good to discuss this as a team, I know it's slithy annoying, but I actually like it as a rule.

File: package.json [hunk 6]
```diff
@@ -33,7 +37,9 @@
     "html-webpack-plugin": "^3.2.0",
     "husky": "^2.3.0",
     "jest": "^24.8.0",
+    "jest-mock-random": "^1.0.2",
     "prettier": "^1.17.1",
+    "ts-jest": "^24.0.2",
```
Comment (phamann) [inline package.json]: Why did you add `ts-jest`?

File: .eslintrc (hunk 5 above)
Comment (psbanka) [reply .eslintrc]: I am open to the discussion. What I find annoying about this rule is that it impacts tights and obvious arrow-functions. For example, it requires something like this:

```
  return values.reduce((memo: number, i: number) => memo + i) / values.length;
```
to be (with prettier):
```
  return (
    values.reduce((memo: number, i: number): number => memo + i) / values.length
  );```

which is not a win, IMHO.

File: package.json (hunk 6 above)
Comment (psbanka) [reply package.json]: because VS-code was upset about `describe` and `it`, etc.

File: .eslintrc (hunk 5 above)
Comment (clight-fastly) [reply .eslintrc]: We have been living without this rule in a separate repo effectively. TS inference is good enough that any issues are typically caught without explicit return type. Adding return type as needed for trickier functions to ensure the body of the function does what the author wants is a nice opt-in. If you use TS `strict` rules, this rule is less impactful.

File: package.json (hunk 6 above)
Comment (psbanka) [reply package.json]: Now that you pointed out the presence of `@types/jest`, I will troubleshoot!

File: .eslintrc (hunk 5 above)
Comment (psbanka) [reply .eslintrc]: will leave this rule out for now and see how much it continues to irritate us (by which I mean "me" 😉)

File: src/util/chooseRenormalize.ts (hunk 2 above)
Comment (psbanka) [reply src/util/chooseRenormalize.ts]: I will sum the probability values.

File: src/util/chooseRenormalize.ts (hunk 1 above)
Comment (psbanka) [reply src/util/chooseRenormalize.ts]: Sounds like this is part of a larger discussion

Comment (psbanka) [reply src/util/chooseRenormalize.ts]: from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
`Math.random()`:
>  The implementation selects the initial seed to the random number generation algorithm; it cannot be chosen or reset by the user.

There is another project that could be used to create a seeded PRNG: https://github.com/davidbau/seedrandom
